packages/tldraw/src/lib/tools/SelectTool/DragAndDropManager.ts

```diff
@@ -44,8 +44,40 @@ export class DragAndDropManager {
 
 		const nextDroppingShapeId = this.editor.getDroppingOverShape(point, movingShapes)?.id ?? null
 
-		// is the next dropping shape id different than the last one?
+		// is the next dropping shape id same as the last one?
 		if (nextDroppingShapeId === this.prevDroppingShapeId) {
+			// Check if all shapes are outside their parents
+			const areAllShapesOutsideTheirParents = movingShapes.every((shape) => {
+				const parent = this.editor.getShape(shape.parentId)
+				if (!parent) return false
+				const parentBounds = this.editor.getShapePageBounds(parent)
+				const shapeBounds = this.editor.getShapePageBounds(shape)
+				if (!parentBounds || !shapeBounds) return false
+				return !parentBounds.includes(shapeBounds)
+			})
+
+			// If all shapes are outside their parents, reparent them to the next dropping shape
+			if (areAllShapesOutsideTheirParents) {
+				const { prevDroppingShapeId } = this
+				const prevDroppingShape = prevDroppingShapeId && this.editor.getShape(prevDroppingShapeId)
+				if (prevDroppingShape) {
+					this.editor
+						.getShapeUtil(prevDroppingShape)
+						.onDragShapesOut?.(prevDroppingShape, movingShapes)
+				}
+
+				for (const shape of movingShapes) {
+					this.editor.reparentShapes(
+						[shape.id],
+						nextDroppingShapeId ?? this.editor.getCurrentPageId()
```

Comment:
hmm, I guess this shouldn't be hardcoded here really. it should probably call `onDragShapesOut` instead.